File: apps/api/src/application/imports/parse-csv.ts
import {
  importColumns,
  requiredImportColumns,
  type ImportCells,
  type ImportColumn,
} from '@ecommerce/shared'
import { parse } from 'csv-parse/sync'

import { InvalidImportFileError } from '@/domain/shared/domain-error'

export interface ParsedRecord {
  readonly line: number
  readonly cells: ImportCells
  readonly blank: boolean
}

export interface ParsedCsv {
  readonly columns: ImportColumn[]
  readonly records: ParsedRecord[]
}

const knownColumns = new Set<string>(importColumns)

const isImportColumn = (name: string): name is ImportColumn => knownColumns.has(name)

const normalizeHeader = (header: string) => header.trim().toLowerCase()

const newlines = (value: string) => value.split('\n').length - 1

type RawRecord = { info: { lines: number }; record: Record<string, string> }

export function parseCsv(content: Buffer | string): ParsedCsv {
  let columns: ImportColumn[] = []
  let raw: RawRecord[]
  try {
    raw = parse(content, {
      bom: true,
      relax_column_count: true,
      skip_empty_lines: false,
      info: true,
      columns: (header: string[]) => {
        columns = header.map(normalizeHeader).filter(isImportColumn)
        return header.map(normalizeHeader).map((name) => (isImportColumn(name) ? name : undefined))
      },
    }) as RawRecord[]
  } catch (error) {
    throw new InvalidImportFileError(
      `The file is not valid CSV: ${error instanceof Error ? error.message : String(error)}`,
    )
  }

  const missing = requiredImportColumns.filter((name) => !columns.includes(name))
  if (missing.length > 0) {
    throw new InvalidImportFileError(`Missing required columns: ${missing.join(', ')}`, missing)
  }
  if (raw.length === 0) {
    throw new InvalidImportFileError('The file has no data rows')
  }

  const records = raw.map(({ info, record }) => {
    const cells: ImportCells = {}
    let blank = true
    let embeddedLines = 0
    for (const column of columns) {
      const value = record[column] ?? ''
      cells[column] = value
      if (value.trim() !== '') blank = false
      embeddedLines += newlines(value)
    }
    return { line: info.lines - embeddedLines, cells, blank }
  })

  return { columns, records }
}

File: apps/api/src/ecommerce_api/application/imports/parse_csv.py
import csv
import io
from dataclasses import dataclass

from ecommerce_api.application.schemas.import_row import (
    IMPORT_COLUMNS,
    REQUIRED_IMPORT_COLUMNS,
    ImportCells,
    ImportColumn,
)
from ecommerce_api.domain.errors import InvalidImportFileError


@dataclass(frozen=True, slots=True)
class ParsedRecord:
    line: int
    cells: ImportCells
    blank: bool


@dataclass(frozen=True, slots=True)
class ParsedCsv:
    columns: list[ImportColumn]
    records: list[ParsedRecord]


def _decode(content: bytes) -> str:
    try:
        return content.decode('utf-8-sig')
    except UnicodeDecodeError as error:
        raise InvalidImportFileError(
            f'The file is not valid CSV: it is not UTF-8 text ({error.reason})'
        ) from error


def _read_rows(text: str) -> list[tuple[int, list[str]]]:
    reader = csv.reader(io.StringIO(text, newline=''), strict=True)
    try:
        return [(reader.line_num, row) for row in reader]
    except csv.Error as error:
        raise InvalidImportFileError(f'The file is not valid CSV: {error}') from error


def parse_csv(content: bytes) -> ParsedCsv:
    rows = _read_rows(_decode(content))
    header = [name.strip().lower() for name in rows[0][1]] if rows else []
    positions: dict[ImportColumn, int] = {
        name: index for index, name in enumerate(header) if name in IMPORT_COLUMNS
    }
    columns: list[ImportColumn] = [name for name in header if name in IMPORT_COLUMNS]

    missing = [name for name in REQUIRED_IMPORT_COLUMNS if name not in positions]
    if missing:
        raise InvalidImportFileError(f'Missing required columns: {", ".join(missing)}', missing)
    if len(rows) < 2:
        raise InvalidImportFileError('The file has no data rows')

    records = []
    for last_line, row in rows[1:]:
        cells: dict[ImportColumn, str | None] = {}
        embedded_lines = 0
        for column, index in positions.items():
            value = row[index] if index < len(row) else ''
            cells[column] = value
            embedded_lines += value.count('\n')
        blank = all(not (value or '').strip() for value in cells.values())
        records.append(ParsedRecord(line=last_line - embedded_lines, cells=cells, blank=blank))
    return ParsedCsv(columns=columns, records=records)
